### PR TITLE
Attempting to enable chunking for CTC (might be not viable).

### PR DESCRIPTION
**Pull request by Narsil**
Co-Authored-By: Sam Waterbury <sam.waterbury@scale.com>

# What does this PR do?

Attempts to recreate a different version of https://github.com/huggingface/transformers/pull/18949

I am under the impression, that the overall approach is doomed to not work because of how mctc models work.

However this PR also enables nice to have features.

This attempts to create the condition for MCTC models (like https://huggingface.co/speechbrain/m-ctc-t-large) 
to work with the `chunk_length_s` argument.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/auto/tokenization_auto.py [hunk 1]
```diff
@@ -166,6 +166,7 @@
                     "MBart50TokenizerFast" if is_tokenizers_available() else None,
                 ),
             ),
+            ("mctct", ("Wav2Vec2CTCTokenizer", None)),
```
Comment (Narsil) [inline src/transformers/models/auto/tokenization_auto.py:169]: This enables the pipeline to automatically load the tokenizer for you.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -57,11 +57,15 @@ def chunk_iter(inputs, feature_extractor, chunk_len, stride_left, stride_right):
         # add start and end paddings to the chunk
         chunk = inputs[i : i + chunk_len]
         processed = feature_extractor(chunk, sampling_rate=feature_extractor.sampling_rate, return_tensors="pt")
+
+        _main = chunk.shape[0]
         _stride_left = 0 if i == 0 else stride_left
         is_last = i + step + stride_left >= inputs_len
         _stride_right = 0 if is_last else stride_right
+        if "input_features" in processed:
+            processed["input_to_features_ratio"] = _main / processed["input_features"].shape[1]
```
Comment (Narsil) [inline src/transformers/pipelines/automatic_speech_recognition.py:66]: Since in that case we are already reducing the size of the sequence length, we start saving the ratio here.

It could probably be calculated from feature_extractor directly but the calculation is likely to be non trivial given all the options I have seen.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 3]
```diff
@@ -258,7 +262,9 @@ def preprocess(self, inputs, chunk_length_s=0, stride_length_s=None):
             # XXX: Carefuly, this variable will not exist in `seq2seq` setting.
             # Currently chunking is not possible at this level for `seq2seq` so
             # it's ok.
-            align_to = self.model.config.inputs_to_logits_ratio
+            # If this variable is not available we cannot know how to align
+            # properly. timestamps are likely to be wrong.
+            align_to = getattr(self.model.config, "inputs_to_logits_ratio", 1)
```
Comment (Narsil) [inline src/transformers/pipelines/automatic_speech_recognition.py:267]: `align_to` is really important for wav2vec2&co since we have X input data values for 1 logits output value, so we **must** align if we don't want to have spurrious replication of letters because we chunked WITHIN a wav2vec2 chunk.

Here since everything is already muddled, it's OK to say 1, but we're also loosing that property.
Since the muddling in `feature_extractor` is likely to be bigger, we're already going to have more issues when stiching back together than this (here we're talking outputting "woorld" instead of "world" which is incorrect but might be acceptable).

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 4]
```diff
@@ -325,11 +335,31 @@ def _forward(self, model_inputs):
                 # Send stride to `postprocess`.
                 # it needs to be handled there where
                 # the pieces are to be concatenated.
-                ratio = 1 / self.model.config.inputs_to_logits_ratio
+                if input_to_features_ratio is None:
+                    ratio = self.model.config.inputs_to_logits_ratio
+                else:
+                    new_ratio = inputs.shape[1] / logits.shape[1]
+                    if isinstance(input_to_features_ratio, list):
```
Comment (Narsil) [inline src/transformers/pipelines/automatic_speech_recognition.py:342]: Every single element of the batch will use the same ratio, and that's guaranteed.

The reason is that we're using the same `feature_extractor` (first ratio) and the same model (second ratio).
The exact numbers might vary because the size of the chunks could vary and not be aligned with said ratio (like 74 400 samples vs 74 350 sampling points).

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 5]
```diff
@@ -325,11 +335,31 @@ def _forward(self, model_inputs):
...
+                            # This shouldn't trigger since ratios should be consistent
+                            # through the pipeline, however since everything is calculated
+                            # a check shouldn't hurt
+                            raise Exception("The ratios are not similar enough")
+                        ratio = input_to_features_ratio[0] * new_ratio
+                        outratio = [ratio] * batch_size
+                    else:
+                        ratio = input_to_features_ratio * new_ratio
+                        outratio = ratio
+                    out["ratio"] = outratio
```
Comment (Narsil) [inline src/transformers/pipelines/automatic_speech_recognition.py:355]: This code is overly complex, but given the end results seems odd, I didn't refactor it yet into something cleaner.

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 6]
```diff
@@ -377,6 +379,49 @@ def test_chunking_fast(self):
...
+                "chunks": [
+                    {"text": "A", "timestamp": (0.3, 0.33)},
+                    {"text": "man", "timestamp": (0.51, 0.72)},
+                    {"text": "said", "timestamp": (0.93, 1.14)},
+                    {"text": "to", "timestamp": (1.2, 1.29)},
+                    {"text": "the", "timestamp": (1.38, 1.47)},
+                    {"text": "universe,", "timestamp": (1.59, 2.4)},
+                    {"text": "Sir,", "timestamp": (2.67, 3.15)},
+                    {"text": "I", "timestamp": (3.33, 3.36)},
+                    {"text": "exist.", "timestamp": (3.6, 4.59)},
```
Comment (Narsil) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:410]: Those offsets are roughly correct, and didn't match in the original PR.

Comment (Narsil) [reply tests/pipelines/test_pipelines_automatic_speech_recognition.py:410]: Because this is a trained big model, we can test ALL the outputs.

This wasn't true in the original test it was copied from, because random models that work letter per letter was outputting a LOT of dummy output, so we had to slice it somehow to keep the test readable. (We could also have chunked the input audio,  but under a certain size, all the output was "", and suddenly triggered a ton of output after a certain length).

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 7]
```diff
@@ -377,6 +379,49 @@ def test_chunking_fast(self):
...
+        )
+
+        # And also for multiple batched inputs
+        n_repeats = 2
+        audio_tiled = np.tile(audio, n_repeats)
+        output = speech_recognizer([audio_tiled], batch_size=2)
+        self.assertEqual(
+            # XXX: This is a probable bug
+            output[0]["text"],
+            "A man said to the universe, Sir, I exist, a man said to the universe, Sirxist.",
```
Comment (Narsil) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:422]: This is the bug.
The first chunk goes all the way to the end of the audio, and has the correct string.
It gets chunked out at the last `Sir` because of the striding. This is expected.
The second chunk, gets the correct string `Sir, I exist`. However the striding forces it to cut out `Sir I e` instead of just `Sir`.

This means the stitching is failing. 

I supposed this is all linked to the initial spectrogram, and is **impossible** to fix.

File: src/transformers/models/auto/tokenization_auto.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/auto/tokenization_auto.py:169]: There should be a default tokenizer for the model type anyway, as users might not always have the class properly set in the config/tokenizer config.

So this change should stay regardless of whether we keep the pipeline or not!

File: src/transformers/utils/import_utils.py [hunk 8]
```diff
@@ -246,6 +246,12 @@
 except importlib_metadata.PackageNotFoundError:
     _pyctcdecode_available = False
 
+_kenlm_available = importlib.util.find_spec("kenlm") is not None
+try:
+    _kenlm_version = importlib_metadata.version("kenlm")
+    logger.debug(f"Successfully imported kenlm version {_kenlm_version}")
+except importlib_metadata.PackageNotFoundError:
+    _kenlm_available = False
```
Comment (sgugger) [inline src/transformers/utils/import_utils.py:254]: We don't need the kenlm version anywhere right? I'd just test for `importlib.util.find_spec("kenlm") is not None` in the util function down there.
```suggestion
```

File: src/transformers/utils/import_utils.py [hunk 9]
```diff
@@ -278,6 +284,10 @@ def is_pyctcdecode_available():
     return _pyctcdecode_available
 
 
+def is_kenlm_available():
+    return _kenlm_available
```
Comment (sgugger) [inline src/transformers/utils/import_utils.py:288]:
```suggestion
    return importlib.util.find_spec("kenlm") is not None
```

**Review by sgugger — APPROVED** [review]
LGTM, but @patrickvonplaten is really the expert on Speech :-)
Thanks!

File: src/transformers/utils/import_utils.py (hunk 8 above)
Comment (Narsil) [reply src/transformers/utils/import_utils.py]: Thanks !

**Review by patrickvonplaten — APPROVED** [review]
Fine to merge for me. 

More generalyl, MCTCT is not used a lot - the more important part here is to make sure that other speech models don't break. Since this is guaranteed, this PR looks good to me!

File: src/transformers/pipelines/automatic_speech_recognition.py (hunk 5 above)
Comment (Narsil) [reply src/transformers/pipelines/automatic_speech_recognition.py:355]: It works as expected, meaning I'm cutting where I should.
But the *results* however (the reconstructed string) is lacking part of the stride.

That's because we cannot attribute produced letters to specific logits, so the cutting logic is unable to know how to reconstruct the string. I don't think it's a property of the mode.

Comment (patrickvonplaten) [reply src/transformers/pipelines/automatic_speech_recognition.py:355]: That's fine for me honestly - M-CTCT is not used a lot at all
